crates/typst/src/layout/grid/layout.rs

```diff
@@ -278,6 +371,83 @@ impl CellGrid {
     }
 }
 
+/// Given a cell's requested x and y, the vector with the resolved cell
+/// positions, the `auto_index` counter (determines the position of the next
+/// `(auto, auto)` cell) and the amount of columns in the grid, returns the
+/// final index of this cell in the vector of resolved cells.
+fn resolve_cell_position(
+    cell_x: Smart<usize>,
+    cell_y: Smart<usize>,
+    resolved_cells: &[Option<Cell>],
+    auto_index: &mut usize,
+    columns: usize,
+) -> HintedStrResult<usize> {
+    // Translates a (x, y) position to the equivalent index in the final cell vector.
+    let cell_index = |x, y| y * columns + x;
```

Comment:
Should we use `checked_mul` and `checked_add` here? I was going to do so, but was kinda afraid that this could hurt performance more than it should depending on the amount of cells in the table... maybe it's not too bad?

(Same applies to e.g. replacing `i + next_auto_index` below with `i.checked_add(next_auto_index)` or `*auto_index = resolved_index + 1` with `*auto_index = resolved_index.saturating_add(1)` for example)

Reply:
Actually, I think it shouldn't be necessary for the auto_index ones, since for that to overflow you'd need `usize::MAX` cells in your table, and you'd probably run out of memory long before that; if not, you'd try to place a cell in an existing position and fail, so whatever. And I think we should optimize for the cases without arbitrary cell positioning (the vast majority of cases), so that decision could make sense.

OTOH we could just accept the performance hit when arbitrary positioning is used, since it's too easy to screw it up (e.g. add several columns to the table and place a cell at a very high `y`, enough to overflow and have it be placed at some random position possibly without error). Not 100% sure if my logic is correct here though.

Reply:
Yeah, just checked... you can produce an overflow with

```typ
#grid(
  columns: 3,
  rows: 2em,
  fill: (x, y) => if calc.odd(x + y) { red.lighten(50%) } else { green },
  grid.cell(y: 6148914691236517207)[a],
)
```

leading to

![cell appears in the second row](https://github.com/typst/typst/assets/9021226/9d8a7d27-1149-407d-9b82-4d93011ebc6b)

so yeah, we definitely need something here. Just probably not for the auto index case.

Reply:
Okay, there we go. Added some checks (which shouldn't impact the `auto_index` case) and tests. :+1:
This is probably solved then.

(Though maybe I should use `*auto_index = next_auto_index.wrapping_add(1)` in case people are using debug mode or something to ensure it won't panic... but eh, maybe the paranoia isn't worth it...)

Reply:
I spotted another problem: this will cause a capacity overflow on `.extend` inside the grid building loop.

```typ
#grid(
  columns: 3,
  rows: 2em,
  fill: (x, y) => if calc.odd(x + y) { red.lighten(50%) } else { green },
  grid.cell(x: 2, y: 6148914691236517204)[a],
)
```

It passes the cell index checks (the cell indices are indeed valid), but I assume the vector is trying to reserve some more space, but it can't.

I guess we could use `reserve_exact` before the `extend`, but is it even worth it? Reserving more space is useful in general anyway. Maybe we could just let this one slip.

Reply:
I think the current overflow checks should be enough to at least avoid unexpected behavior. Crashes are fine imo. Marking this as resolved.